feat(docs): equality semantics (#2923)
winglang/wing: docs/03-language-guide/14-equality.md

---
title: Equality
id: equality
description: A reference guide to equality in Wing.

> keywords: [Winglang equality, Wing equality, equals, identity]

Reviewer: I don't think we need to include `wing` in our keywords. Ideally, we should find a way to inject it for every page in the docs automatically. or maybe it's just not needed at all.

```suggestion
keywords: [equality, equals, identity]
```

---

> Checking for equality is performed with the `==` operator in Wing. It returns `true` if the two values are equal, and `false` otherwise.

Reviewer: nit: I feel like our reference docs should avoid mentioning "in Wing" unless it's ambiguous (e.g. when comparing to other languages)

```suggestion
Checking for equality is performed with the `==` operator. It returns `true` if the two values are equal, and `false` otherwise.
```

Contextually, all of these docs are referring to the world of Wing so it's redundant.

Author: I agree

The main difference between equality in JavaScript and Wing is that `==` in Wing is not allowed to compare values of different types. For example, `1 == "1"` is not allowed in Wing, and will result in a compile-time error.

> The execution phase ([preflight or inflight](../02-core-concepts/01-preflight-and-inflight.md)) that a value was created in does not affect its equality. For example, a value created in preflight can be equal to a value created in inflight.

Reviewer: This weirdly feels like it's too complicated to be correct, but I can't think of any issues, so 👍

Author: Yeah, it's a little optimistic, for sure :-)

One question I had was if you have two instances of a class created in preflight, and then you lift them into inflight, can you still safely check them for equality? In any case, I'm sure we'll figure out in due time

Reviewer 2: I feel like this paragraph is a pretty advanced statement and should likely go at the bottom of this section (perhaps also the one about JavaScript).

Equality in Wing is a symmetric and transitive relationship - that is, (1) if `a == b`, then `b == a`, and (2) if `a == b` and `b == c`, then `a == c`.

The following is a set of rules for checking equality:

## Basic types

1. Two `str` values are equal if they have the same characters in the same order.

> 2. Two `num` values are equal if they have the same floating-point value. `NaN` is not equal to any value, including itself.

Reviewer: Should we reference https://en.wikipedia.org/wiki/IEEE_754? Assuming we're gonna go by that (as js does).
That would include specifying that -0 == +0.

  CHEESE,
  PINEAPPLE,
}

assert(PizzaTopping.CHEESE == PizzaTopping.CHEESE);
assert(PizzaTopping.CHEESE != PizzaTopping.PINEAPPLE);
```

## Classes and interfaces

Two class instances or interface-satisfying objects are equal if they are the same instance. This means that two class instances, or interface-satisfying objects that have the same data are not necessarily equal, since they may have been created in different places.

```js
class Shop {
  hats: num;
  init(hats: num) {
    this.hats = hats;
  }
}

let shop1 = new Shop(1);
let shop2 = new Shop(1);
let shop3 = shop1;

assert(shop1 != shop2);
assert(shop1 == shop3);
```

## `Json`

Two `Json` values are equal if they contain the same structure and values. The following rules apply:

1. Two `Json` values are equal if they are both `null`.
2. Two `Json` values are equal if they are both `bool` values and are equal.
3. Two `Json` values are equal if they are both `num` values and are equal.
4. Two `Json` values are equal if they are both `str` values and are equal.
5. Two `Json` values are equal if they are both `Array` values and are equal.
6. Two `Json` values are equal if they are both `Map` values and are equal.

```js

> assert(Json.null == Json.null); // TODO: https://github.com/winglang/wing/issues/1819

Reviewer: I know the syntax isn't actually finalized, but given we added nil I feel like this is also a possible alternative.

```suggestion
assert(Json nil == Json nil); // TODO: https://github.com/winglang/wing/issues/1819
```

Author: Sure, let's try it

---
title: Equality
id: equality
description: A reference guide to equality in Wing.
keywords: [equality, equals, identity]
---

Checking for equality is performed with the `==` operator. It returns `true` if the two values are equal, and `false` otherwise.

> The main difference between equality in JavaScript and Wing is that `==` in Wing is not allowed to compare values of different types. For example, `1 == "1"` is not allowed in Wing, and will result in a compile-time error.

Reviewer 2: ```suggestion
> The main difference between equality in JavaScript and Wing is that `==` in Wing is not allowed to compare values of different types. For example, `1 == "1"` is not allowed in Wing, and will result in a compile-time error.
```

The execution phase ([preflight or inflight](../02-core-concepts/01-preflight-and-inflight.md)) that a value was created in does not affect its equality. For example, a value created in preflight can be equal to a value created in inflight.

Equality in Wing is a symmetric and transitive relationship - that is, (1) if `a == b`, then `b == a`, and (2) if `a == b` and `b == c`, then `a == c`.

The following is a set of rules for checking equality:

> 

Reviewer 2: I think it's worth defining the terms "by value" and "by reference" here and using the at the beginning of each section.

Also worth mentioning the concept of deep versus shallow comparison.

Author: Good idea, I think adding "by value" / "by reference" improves the explanation 👍

the way I've defined the semantics hopefully doesn't require burdening users with understanding the difference between deep and shallow comparisons (the suggested direction is if we're checking if two arrays are equal, we perform `==` on every element recursively); but if that doesn't end up working, then we can add it back.

## Basic types

> 

Reviewer 2: ```suggestion

Basic types are compared *by value*.

```

1. Two `str` values are equal if they have the same characters in the same order.

> 2. Two `num` values are equal if they have the same floating-point value. The [IEEE 754] standard is used for storing numbers, which means that for example `-0 == +0`. `NaN` is not equal to any value, including itself.

Reviewer 2: with floating point numbers, it is sometimes desirable to compare them with some possible accuracy level (but that's likely a `math.equal()` thing).

Author: Good point, I think we will need a math utility for that use case. `==` for numbers is still sensible for integers of course

3. Two `bool` values are equal if they are both `true` or both `false`.
4. Two `duration` values are equal if they have the same number of milliseconds.

> 5. Two `T?` types (optional `T` values) are equal if they are both empty (`nil`) or both non-empty, and if they are both non-empty, their inner values are equal. A value of type `T?` is never equal to a value of type `T`.

Reviewer 2: Something that came up [here](https://github.com/winglang/wing/pull/2918#discussion_r1230673067) is that maybe we should allow comparing `T?` and `T`:

```js
let var x: str? = "hello";

if x == "hello" {
  log("x is hello");
}

x = nil;
if x != "hello" {
  log("x is not hello");
}
```

Reviewer 3: I'm for staying explicit about this for now. But I also see the value (pun intended) of implicitly checking for nil.
If we take that course there's the question of being recursive about it. Can we compare a `T???` with a `T`? Can we compare a `T??` with a `T?`? I think we should allow all of these if we allow any of them.

Reviewer 2: > I'm for staying explicit about this for now.

What's the motivation?

Author: > What's the motivation?

I think the idea is to require unwrapping optional values, no?

[IEEE 754]: https://en.wikipedia.org/wiki/IEEE_754

## Collection types

> Wing contains six collection types: `Array`, `MutArray`, `Map`, `MutMap`, `Set`, and `MutSet`.The following rules apply to all of them:

Reviewer 2: 👍

1. Two collections are equal if they have the same number of elements, and if each element in the first collection is equal to the corresponding element in the second collection. (The order of elements only matters for `Array` and `MutArray`.)
2. The mutability of a collection does not affect its equality. In other words, a `MutArray` is equal to an `Array` with the same elements, and a `MutMap` is equal to a `Map` with the same keys and values.
3. Only collections of the same "kind" can be equal. For example, an `Array` cannot be equal to a `Map`, and a `MutArray` cannot be equal to a `MutMap`.

> 

Reviewer 2: This is currently not the behavior correct? Maybe worth mentioning and adding a tracking issue.

Author: I've added links to issues covering aspects of equality that have not been implemented yet.

## Function types

> Two functions are equal if they are both the same function. This means that two functions that have the same code are not necessarily equal, since they may have been defined in different places.

Reviewer 2: ```suggestion
Two functions are equal if they are both the same object (by reference). This means that two functions that have the same code are not necessarily equal, since they may have been defined in different places.
```

Reviewer 3: Right, this is a special case of non-primitive equality (always by reference).

> Two functions are equal if they are both the same function. This means that two functions that have the same code are not necessarily equal, since they may have been defined in different places.

Author: Good call, added.

```js
let f1 = (x: num): num => x + 1;
let f2 = (x: num): num => x + 1;
let f3 = f1;

assert(f1 != f2);

> assert(f1 == f3);

Author: Aha, copilot got me there... thanks for catching this

```

## Enums

Two enum values are equal if they refer to the same enum case.

```js
enum PizzaTopping {
  CHEESE,
  PINEAPPLE,
}

> assert(PizzaTopping.CHEESE == PizzaTopping.CHEESE);
> assert(PizzaTopping.CHEESE != PizzaTopping.PINEAPPLE);

Reviewer 3: ```suggestion
assert(PizzaTopping.CHEESE == PizzaTopping.CHEESE);
assert(PizzaTopping.CHEESE != PizzaTopping.PINEAPPLE);
let weird_taste = PizzaTopping.PINEAPPLE;
assert(weird_taste == PizzaTopping.PINEAPPLE);
```

Author: Hey, what are you trying to say about pineapple pizza...? 😜

```

> ## Classes and interfaces

Author: Right I could imagine users might want a way to define / override equality for custom types. Let's open a separate issue for this

> Two class instances or interface-satisfying objects are equal if they are the same instance. This means that two class instances, or interface-satisfying objects that have the same data are not necessarily equal, since they may have been created in different places.

Reviewer 2: ```suggestion
Two class instances or interface-satisfying objects are equal if they are the same instance (by reference). This means that two class instances, or interface-satisfying objects that have the same data are not necessarily equal, since they may have been created in different places.
```

```js
class Shop {
  hats: num;
  init(hats: num) {
    this.hats = hats;
  }
}

let shop1 = new Shop(1);
let shop2 = new Shop(1);
let shop3 = shop1;

assert(shop1 != shop2);
assert(shop1 == shop3);
```

## `Json`

> Two `Json` values are equal if they contain the same structure and values. The following rules apply:

Reviewer 2: ```suggestion
Two `Json` values are equal if they contain the same structure and values (deep by reference equality). Another way to think about it is the two `Json` values are equal if their stringified representation is equal.

The following rules apply:
```

Reviewer 3: why "deep reference equality"? just "deep equality". References seem not relevant for how we want to do `Json` compares.

> 1. Two `Json` values are equal if they are both `null`.

Author: Sure

2. Two `Json` values are equal if they are both `bool` values and are equal.
3. Two `Json` values are equal if they are both `num` values and are equal.
4. Two `Json` values are equal if they are both `str` values and are equal.
5. Two `Json` values are equal if they are both `Array` values and are equal.
6. Two `Json` values are equal if they are both `Map` values and are equal.

```js
assert(Json nil == Json nil); // TODO: https://github.com/winglang/wing/issues/1819
assert(Json true == Json true);
assert(Json false == Json false);
assert(Json 1 == Json 1);
assert(Json -0.42 == Json -0.42);
assert(Json "foo" == Json "foo");
assert(Json [1, 2, 3] == Json [1, 2, 3]);
assert(Json { "foo": 1, "bar": 2 } == Json { "foo": 1, "bar": 2 });
```

## Structs

> Two struct values are equal if they have the same type and all of their fields are equal. The order in which fields are defined does not matter.

Reviewer 2: ```suggestion
Two struct values are equal if they have the same type and all of their fields are equal (based on rules of equality of their type). The order in which fields are defined does not matter.
```

Reviewer 3: The comment about the order sees a bit misleading to me. If they are the same type they are the same type.

Author: I guess the question how does someone know that `Foo { a: 1, b: 2 }` is the same as `Foo { b: 2, a: 1 }`? It sounds like you're saying that's already part of the definition of a struct, in which case I'm ok taking that sentence out.

```js

> struct Node {

Reviewer 2: That's a pretty esoteric example...

Author: Changed it a bit
